Fixes #7957: Convert e2e bash script to python script

cuichenli opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
One new `run_e2e_tests.py` script is created.
*Note*:
- I can successfully start the e2e tests on Ubuntu and Windows. But I have not tried to run all the tests yet. I will later create a GitHub action to run the tests on both Windows and Ubuntu. 

- There are a few changes I pulled from the #8139 , as I need `build.py` to run on Windows. 

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

cuichenli committed:
Add .env file to git ignore

cuichenli committed:
Introduce psutil

cuichenli committed:
Setup parts

cuichenli committed:
Start google app engine

cuichenli committed:
Actual test part

cuichenli committed:
Fix minor issues

Commit:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

Commit:
Add except for acessdeny

Commit:
Tweak webdriver manager and apply common variables

Commit:
Fix webpack command and tsc command

Commit:
Fix different kinds of windows path issues

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into fix-build-script-on-windows

cuichenli committed:
Fix linter issue

cuichenli committed:
Use os.path variables and functions

cuichenli committed:
Replace normalize function with built-in ones.

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into fix-build-script-on-windows

cuichenli committed:
Revert back to customized convert function, as built-in one is not working as intended.

Commit:
Merge branch 'fix-build-script-on-windows' of github.com:cuichenli/oppia into convert-e2e-to-python

Commit:
Fix linter issues

Commit:
Add docstrings

Commit:
Fix check architecture

Commit:
Fix some minor issues

cuichenli committed:
Add a few unit tests

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Introduce swap_with_checks

cuichenli committed:
Introduce inplace_replace

cuichenli committed:
Add and change a few tests

cuichenli committed:
Improve swap_with_checks

cuichenli committed:
Finish the unit tests

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into fix-build-script-on-windows

cuichenli committed:
Rename function and variables

cuichenli committed:
Unit tests for common

cuichenli committed:
All unit tests for run_e2e_tests

cuichenli committed:
Add unittest for non-ascii code

cuichenli committed:
Improve unittests

cuichenli committed:
Refine clean up

cuichenli committed:
Rename function and apply it in a few more places.

cuichenli committed:
Add unit test for convert_to_posixpath

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into fix-build-script-on-windows

cuichenli committed:
Fix linter issue

cuichenli committed:
Merge branch 'fix-build-script-on-windows' of github.com:cuichenli/oppia into convert-e2e-to-python

cuichenli committed:
Add cleanup unittest

cuichenli committed:
Add message to swap_with_checks

cuichenli committed:
Broaden psutil exception

cuichenli committed:
Use current python executable in subprocess

cuichenli committed:
Remove non-ascii code in unit test, as pylint can not recognize

cuichenli committed:
Add docstring, module string, a little change to the format

cuichenli committed:
Fix some linter issues and improve the unit test structure

cuichenli committed:
Fix linter issue

cuichenli committed:
Add codeowner

cuichenli committed:
Use constant for version

cuichenli commented on .gitignore:
```diff
@@ -35,6 +35,7 @@ venv/
 yarn-error.log
 oppia-release-credentials.json
 sheets.googleapis.com-python.json
+.env
```
Add `.env` so I can use the pylint extension on vscode. Otherwise, the python program has no idea where to find the packages.

cuichenli commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,440 @@
+    run_webdriver_manager(
+        ['update', '--versions.chrome', CHROME_DRIVER_VERSION])
+    run_webdriver_manager(
+        ['start', '--versions.chrome', CHROME_DRIVER_VERSION,
```
There were two `start` on the `bash` script, but it looks like the second one is not working. I tried to run the two `start` manually and the second one was throwing an error that the port 4444 was already bound.

cuichenli commented:
Two strange things happened....
The backend test failed to detect `run_e2e_tests_test.py` and also failed to import `common_test.py`. 
The linter is reporting wrong import order while when I run the `pre_commit_linter` locally, there was no such an error. I tried `isort` locally and the result is as the same as the current sequence. 
Did I miss something?

kevinlee12 commented:
For the backend test, I think this was something Luis also ran into. See https://github.com/oppia/oppia/pull/7772. 

As for the linter test, I was able to get it locally. It's this [line](https://github.com/oppia/oppia/pull/8159/files#diff-9a8d10ee24e59bf752a5944de2f54823R37) that it's complaining about.

cuichenli committed:
Change base class. Change sort sequence

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli commented:
I changed the import sequence when I push I got :
```
ERROR: /home/cuichen/opensource/oppia/scripts/common_test.py Imports are incorrectly sorted.
--- /home/cuichen/opensource/oppia/scripts/common_test.py:before        2019-12-10 21:20:27.526863
+++ /home/cuichen/opensource/oppia/scripts/common_test.py:after 2019-12-11 09:06:18.358033
@@ -30,11 +30,10 @@
 import tempfile

 from core.tests import test_utils
-
-import psutil
 import python_utils
 import release_constants

+import psutil

 from . import common



FAILED   Import order checks failed, file imports should be alphabetized, see affect files above.
```
Anyway, I just bypassed it. 

Noticed one thing, the `isort` module in `pre_commit_linter` is not imported from `oppia_tools`, instead, it is imported from `pylint`, is that intended?  

Also fixed that `run_backend_test` is not picking up the new test, but I have no idea what happened to `common_test.py` in the previous build.

seanlip commented on .github/CODEOWNERS:
```diff
@@ -132,6 +132,8 @@
 /scripts/release_scripts/deploy_test.py @ankita24079 @kevinlee12
 /scripts/run_backend_tests.py @kevinlee12
 /scripts/run_e2e_tests.sh @kevinlee12
```
This should be deleted, right?

cuichenli replied:
This one was outdated.

kevinlee12 commented:
> Noticed one thing, the isort module in pre_commit_linter is not imported from oppia_tools, instead, it is imported from pylint, is that intended?

Hmm, where do you see that?

cuichenli commented:
Noticed that there is no isort path in https://github.com/oppia/oppia/blob/c30455f5258a36061a6889dcb6db5f261d92c5eb/scripts/pre_commit_linter.py#L553 
Then I print the path explicitly and found the path is from `pylint`.

kevinlee12 commented:
Ah, in that case, it's probably intentional.

cuichenli commented:
Feel that isort module in oppia_tools is never used...

kevinlee12 commented:
Go ahead and file an issue looking into removing that (I don't want to expand the scope of this PR).

cuichenli committed:
Add path to gae_suite.py

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Fix code coverage

kevinlee12 reviewed (commented):
What you have there looks great, though I had two comments:
- Remove the old shell script for e2e
- Change the .travis.yml file so that it points to the e2e python script.

seanlip commented:
Hi @cuichenli -- just a heads-up that I'm kinda waiting for https://github.com/oppia/oppia/pull/8139 to get merged before looking at this. Can you please reassign me once that happens?

Thanks!

cuichenli committed:
Get exit code and skip installation

cuichenli committed:
Use python script in travis

cuichenli committed:
Merge branch 'develop' of https://github.com/oppia/oppia into convert-e2e-to-python

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Add bak file to gitignore

cuichenli committed:
Refine inplace replace

cuichenli committed:
Add support for community_dashboard

cuichenli committed:
Remove run_e2e_tests.sh

cuichenli committed:
Merge branch 'convert-e2e-to-python' of github.com:cuichenli/oppia into convert-e2e-to-python

cuichenli committed:
Fix linter issue

kevinlee12 reviewed (approved):
lgtm!

kevinlee12 commented:
@cuichenli there is a linter error regarding a code owner file, ptal!

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Remove run_e2e_tests.sh in codeowner file

cuichenli committed:
Add return types to docstring

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Fix linter issue

kevinlee12 commented:
@cuichenli is this ready for other reviewers to look?

cuichenli commented:
> @cuichenli is this ready for other reviewers to look?

@kevinlee12 I am not sure, in theory, the code is ready but apparently there are two e2e tests keep failing. I don't know it is due to my changes or something else.

kevinlee12 commented:
Let's restart it and see what happens. Does it work locally?

kevinlee12 commented:
Can you also merge from develop and try again? (I'm seeing some older stuff in here).

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli commented:
@kevinlee12 Just merged. 

I have tried those tests on my local machine (in a ubuntu 16.04 docker container, with headless chrome) and they are working correctly.

seanlip commented:
Hi @cuichenli, do you think you could try debugging with a local version of Travis (as described [here](https://github.com/oppia/oppia/wiki/Setup-your-own-Travis-instance))? You can temporarily comment out the other tests so that the development cycle is quicker, but I think you'll need to run things in the actual Travis environment in order to figure out what's going on. We had another report recently (#8268) of skew between the Travis env and local machines.

In particular it may be worth getting to the root of the webpack "exit status 2" error. I searched online a bit for that and found that it's due to stuff being printed to stderr by webpack. Might be worth seeing if you can find a way to get a hold of that log.

cuichenli commented:
@seanlip I will setup another branch to debug it. In the meantime - should I close this PR for now?

kevinlee12 commented:
Yeah, let's do that so we don't crowd the PR list here.

cuichenli committed:
Handle the case where hashes.json is missed

cuichenli committed:
Install chrome on travis

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Replace open with python_utils.open_file

cuichenli committed:
Remove redundant lines

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli commented:
I have fixed the travis issue. 

The backend test failed because the test coverage of `install_chrome_on_travis.py` does not meet the requirements (this file does not have a corresponding unit test). I will raise another PR to fix this issue (or should I fix it in this PR?). In the meantime, the review can start.

kevinlee12 commented:
@cuichenli the tests will have to be done in this PR, otherwise if we merge this as-is, all backend builds will fail.

cuichenli committed:
Fix missed space in travis.yml

cuichenli committed:
Fix install chrome script unit test

seanlip commented on .gitignore:
```diff
@@ -33,6 +34,9 @@ venv/
 .vagrant/*
 .lock
 yarn-error.log
+oppia-release-credentials.json
```
Hm, pretty sure this shouldn't be here. Did you merge from develop correctly? Please double-check the "Files Changed" tab to ensure that the changes are as expected.

cuichenli replied:
I mistakenly added this in a merge. Removed.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -88,6 +89,7 @@
     os.path.join(common.THIRD_PARTY_DIR, 'six-1.12.0'),
     os.path.join(common.THIRD_PARTY_DIR, 'soupsieve-1.9.1'),
     os.path.join(common.THIRD_PARTY_DIR, 'webencodings-0.5.1'),
+    os.path.join(common.THIRD_PARTY_DIR, 'webencodings-0.5.1'),
```
Weird ... why duplicate this line?

cuichenli replied:
Removed.

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -266,6 +266,8 @@ def main(args=None):
         ('browsermob-proxy', '0.8.0', common.OPPIA_TOOLS_DIR),
         ('selenium', '3.13.0', common.OPPIA_TOOLS_DIR),
         ('PyGithub', '1.43.7', common.OPPIA_TOOLS_DIR),
+        ('pygsheets', '2.0.2', common.OPPIA_TOOLS_DIR),
```
Is this addition intentional?

cuichenli replied:
Removed.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+                    function_will_invoke_popen()
+
+        Args:
+            obj: Any. The target that you want to swap a function with.
```
We use `*` to represent `Any`.

Also, shouldn't the description be: `the Python object whose attribute you want to swap`?

cuichenli replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+            self, obj, attr, newvalue, expected_args=None, expected_kwargs=None,
+            called=True, called_times=None,
+            expected_exception=IntendedException):
+        """Swap an object's function value within the context of a
```
This docstring is a bit messy. Format it as: general description, examples, note -- and then have Args and Yields sections, in order, at the end. Separate each pair of adjacent sections with a single newline.

cuichenli replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+            If you want to check subprocess.Popen is invoked twice
+            like `subprocess.Popen(['python'], shell=True)` and
+            `subprocess.Popen(['python2], shell=False), you can first
+            define the mock  function, then the swap, and just run the
```
Drop extra space after "mock".

cuichenli replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+            expected_kwargs: None|dict|list[dict]. The expected key word args
+                you want this function to be invoked with. Similar to
+                expected_args.
+            called: Bool. Whether the function is invoked. This will always be
```
Bool --> bool

cuichenli replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+                popen_swap = self.swap_with_checks(
+                    subprocess, 'Popen', mock_popen, expected_args=[
+                        (['python'],), (['python2'],)], expected_kwargs={
+                            'shell': True,
```
This is weird. Wouldn't the dictionary only keep the last value for the key?

cuichenli replied:
Updated

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+                            'shell': False
+                        })
+                with popen_swap:
+                    function_will_invoke_popen()
```
function_that_invokes_popen()

cuichenli replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+            newvalue: Function. The new function you want to use.
+            expected_args: None|tuple|list[tuple]. The expected args that you
+                want this function to be invoked with. When its value is None,
+                args will  not be checked. If the value type is tuple, means
```
Remove extra space after "will". (There are double spaces in odd places elsewhere, please remove.)

cuichenli replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+            obj: Any. The target that you want to swap a function with.
+            attr: str. The name of the function to be swapped.
+            newvalue: Function. The new function you want to use.
+            expected_args: None|tuple|list[tuple]. The expected args that you
```
I think this argument is doing too much heavy lifting. How about getting rid of "called" and "called_times" and having this and expected_kwargs be lists of tuples/dicts? The number of elements in the list is the number of times the function is called. Each element in the list corresponds to the expected args and kwargs for the corresponding call, respectively. If there are any mismatches then that's an error.

cuichenli replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+                checked.
+            called_times: int|None. How many times this function will be called.
+                If the value is None, this will not be checked.
+            expected_exception: Exceptiono. The exception that you expect to be
```
Exception is misspelled.

It is weird to raise an exception (even an "Intended" one) when doing so isn't actually needed. Could we not have this feature and use self.assertRaisesRegexp etc. where needed?

cuichenli replied:
Removed.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
+                IntendedException.
+
+        Note:
+            Currently the function does not support specify the ordered invoke
```
This is concerning :) See note above for a suggested fix.

cuichenli replied:
Fixed

seanlip commented on scripts/common.py:
```diff
@@ -67,7 +73,7 @@ def is_linux_os():
 def is_x64_architecture():
     """Check if the architecture is on X64."""
-    return ARCHITECTURE == 'x86_64'
+    return ARCHITECTURE.endswith('64')
```
Perhaps explain more in the docstring at least two possible values that warrant this not being an equality check, or add a link to something authoritative that says this is the thing to do.

cuichenli replied:
Changed the implementation and added reference.

seanlip commented on scripts/common.py:
```diff
@@ -422,6 +428,26 @@ def check_prs_for_current_release_are_released(repo):
 
 
+def kill_processes_based_on_regex(pattern):
+    """Kill processes based on provided regex.
```
Kill any processes whose names (?) match the provided regex.

(Please replace "names" with whatever is appropriate.)

cuichenli replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -422,6 +428,26 @@ def check_prs_for_current_release_are_released(repo):
+                python_utils.PRINT('Killing %s ...' % cmdline)
+                process.kill()
+        except psutil.Error:
+            continue
```
Shouldn't we print something, if this case isn't expected to happen? Or will this retry? (If so, please add a comment to indicate that.)

cuichenli replied:
In theory, the exception will not influence the actual behaviour. Exceptions would be threw include `AccessDenied`, `NoSuchProcess`, `ZombieProcess`, `TimeoutExpired`. Those exceptions can just be ignored. I will add a comment.

https://psutil.readthedocs.io/en/latest/#exceptions

seanlip commented on scripts/common.py:
```diff
@@ -437,6 +463,23 @@ def convert_to_posixpath(file_path):
 
 
+def inplace_replace_file(filename, regex_pattern, replace):
+    """Replace the file content inplace with regex pattern.
```
Just to clarify: is the idea that, you are looking at the file line by line, and for each line, you are finding all matches for the regex, and replacing them with the new string?

If so, change the arg `replace` to `replacement_string`, since replace sounds like a function name. Also the docstring should be clearer: "Finds all instances of regex_pattern in the given file and replaces them with replacement_string."

Please also document that the regex only works for stuff that is in the same line of code (i.e. regexes containing newlines aren't going to work -- though I wonder why that is? It would be better not to do things that way, IMO, and process the entire file as a string). Perhaps this is something that could be made part of the function name as well.

Also document what happens if regex finds are overlapping. E.g. file contains "aaaaaaa" and regex is "aa" and replacement is "b". There are 6 instances of "aa" in the file; which of those get replaced?

seanlip replied:
Also if you can simplify this function enough, then you can probably just inline it.

cuichenli replied:
I did not expect this function to achieve such a comprehensive job. This functionality can be found at many places as inline code across the scripts runner code base, so I thought it maybe better to provide a function (also when writing unit test, I just need to mock this function instead of a lot of ones).

As far as I can see, this function in the code base is just doing some trivia jobs, so I did not consider so many cases. If you feel it is better to inline it, I will just inline it.
https://github.com/oppia/oppia/blob/1f4b118c7b69d23439599389537d7db19180979b/scripts/start.py#L106
 https://github.com/oppia/oppia/blob/1f4b118c7b69d23439599389537d7db19180979b/scripts/start.py#L117
https://github.com/oppia/oppia/blob/1f4b118c7b69d23439599389537d7db19180979b/scripts/create_expression_parser.py#L54
https://github.com/oppia/oppia/blob/1f4b118c7b69d23439599389537d7db19180979b/scripts/create_expression_parser.py#L65

seanlip replied:
Ah, thanks for noticing the duplication. I think it's fine to create a utility function if its contract is defined really clearly. In particular, it is OK to stipulate that the input regex is applied to the file line by line (so I retract my parenthetical comment in the third paragraph).

cuichenli replied:
Done

seanlip commented on scripts/common_test.py:
```diff
@@ -557,6 +583,91 @@ def mock_get_label(unused_self, unused_name):
+        expected_lines = [line.replace('\n', '') for line in origin_lines]
+        expected_lines[2] = '"DEV_MODE": true'
+        # pylint: disable=unused-argument
+        def mock_input(files, inplace, backup):
```
Don't mock this. Do the operation on a test file. You can use core/tests/data to store test files and expected values.

Make sure to restore things to their original state after the test.

Also please make sure that the tests are comprehensive, in terms of behaviour. E.g. they test regexes with newlines, overlapping matches, etc.

cuichenli replied:
I tried to put an actual file, but the problem is that there are many  `PRINT` invoked in other context, which are all dumped into the actual file.

seanlip replied:
I think that's a negative consequence of us using the fileinput API. Let's maybe not do that; let's use 

```
    with open(..., 'r') as f:
        ...
```

and

```
    with open(..., 'w') as f:
        ...
```

instead.

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,473 @@
+    screeenshots_dir = os.path.join(os.pardir, 'protractor-screenshots')
+    if not os.path.isdir(screeenshots_dir):
+        return
+    python_utils.PRINT("""
```
The indentation here makes things a bit hard to read. I suggest:

```
    python_utils.PRINT(
        'Note: If ...'
        'core/tests/...'
        '...')
```

Note that all the strings will be concatenated prior to printing, so ensure that you have spaces at the ends of the strings as appropriate. Also note that you do not need commas between the strings.

Ditto elsewhere for other multiline strings.

cuichenli replied:
Done

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,473 @@
+    """Check if exists running oppia instance.
+
+    Args:
+        ports: list[int]. A list of ports that oppia instances may be running.
```
Our format is `list(int)`.

that oppia instances may be --> where Oppia instances might be

cuichenli replied:
Done

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,473 @@
+        ports: list[int]. A list of ports that oppia instances may be running.
+
+    Return:
+        boolean represents whether the ports are open.
```
`bool: Whether the ports are open.` In general please make sure to use the correct typeinfo (it should typically be a python type or a class name) and follow the conventions for writing docstrings in the codebase.

Though, this docstring is confusing -- at least one port, or all ports? Please clarify.

cuichenli replied:
Done

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,473 @@
+        port: int. The port number to wait.
+    """
+    current_time = 0
+    while not common.is_port_open(port) and current_time < WAIT_PORT_TIMEOUT:
```
WAIT_PORT_TIMEOUT should include a unit in the constant.

So should current_time. Though maybe change that to time_waited_so_far_msecs (or similar). "current time" sounds like the "current time of day".

cuichenli replied:
Done

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,473 @@
+
+    Args:
+        feconf_file_path: str. Path to the feconf.py file.
+        enable_community_dashboard: boolean. Represents whether community
```
Here and elsewhere, follow existing docstring conventions. Use `bool`, `list(str)`, and make sure you actually have docstrings for all args. (See functions below as well; please proofread the file in its entirety.)

cuichenli replied:
Done

seanlip reviewed (changes requested):
Thanks, @cuichenli! Took a first pass.

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Remove merge conflict left over

cuichenli committed:
Format the docstrings

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Remove called times, change kwargs and args to list

cuichenli committed:
Change implementation of is_x64_architecture and add reference

cuichenli committed:
Change a few comments in common.py

cuichenli committed:
Change variable names and print format

cuichenli committed:
fix some broken unit tests

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Update circleci config

cuichenli committed:
Fix linter issues

cuichenli committed:
Clarify one docstring

cuichenli committed:
Add no cover to is_x64_architecture

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+        """Swap an object's function value within the context of a
+        'with' statement. The object can be anything that supports
+        getattr and setattr, such as class instances, modules, ...
+        Yields:
```
Add a blank line above this one, since it's a separate section.

In fact -- also move this block to the end of the docstring. Docstring sections have a specific order: general description, args, returns/yields, raises.

cuichenli replied:
Updated

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+            newvalue: Function. The new function you want to use.
+            expected_args: None|list(tuple). The expected args that you
+                want this function to be invoked with. When its value is None,
+                args will not be checked.If the value type is list, the function
```
Add space after the period at the end of "checked".

cuichenli replied:
Updated

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+        Args:
+            obj: *. the Python object whose attribute you want to swap.
+            attr: str. The name of the function to be swapped.
+            newvalue: Function. The new function you want to use.
```
newvalue --> new_value
Function --> function

cuichenli replied:
Updated

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+                args will not be checked.If the value type is list, the function
+                will check whether the called args is the first element in the
+                list. If matched, this tuple will be removed from the list.
+            expected_kwargs: None|list(dict). The expected key word args
```
key word --> keyword

cuichenli replied:
Updated

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+            expected_kwargs: None|list(dict). The expected key word args
+                you want this function to be invoked with. Similar to
+                expected_args.
+            called: bool. Whether the function is invoked. This will always be
```
is --> is expected to be

cuichenli replied:
Updated

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+                checked.
+        """
+        original = getattr(obj, attr)
+        msg = 'Failed in check function %s.%s' % (obj.__name__, attr)
```
Can you make this message clearer? I read it a couple of times and I don't understand it -- i.e. I think I won't understand it as a developer if I were to encounter it. Could you make it clearer what the developer is supposed to do if they see this error?

cuichenli replied:
This message is only designed to tell the user which function failed the validation. 

The detailed information, e.g.: the original value and expected value, is added by the `assert` functions - that is also why I added the `self.longMessage = True`. By default there is no origin and expected value print out by the assert function so I need to specify this value to be `True` to tell it to do so. 

I switched the value back to `False` later on because I do not want to change the other `assert`'s behaviours. 

Here is an example of the actual output:
```
AssertionError: {'args': []} != {u'args': [1, 2, 3]}
- {'args': []}
+ {u'args': [1, 2, 3]}
?  +         +++++++
 : Failed in check function scripts.setup_gae.main
```

seanlip replied:
Ah, ok. That sounds fine to me, maybe add a comment in the code to explain that "this message will be printed after the AssertionError debug logs to explain which function failed the validation".

But, looking at your message, I'm still a bit confused about where to look (as a developer) -- where is the "check function scripts.setup_gae.main"? Will there be a stacktrace that I can use to debug? Also I am not sure what the relevance of scripts.setup_gae.main is, here -- could you explain?

(Maybe we can rephrase it to something like: "Expected checks failed when swapping out scripts.setup_gae.main in tests." -- if that is what it actually means.)

cuichenli replied:
Thanks, that is what I mean. 
There is a stack trace in the error message.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+        """
+        original = getattr(obj, attr)
+        msg = 'Failed in check function %s.%s' % (obj.__name__, attr)
+        def wrapper(*args, **kwargs):
```
For readability, add a new line above and below this internal function.

cuichenli replied:
Updated

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+            # This will show the detailed assert message.
+            self.longMessage = True
+            yield
+        except Exception as exception:
```
I _think_ you can simplify this to "except Exception" (and just use "raise" a few lines down). Will that preserve the stacktrace accordingly?

cuichenli replied:
Updated. This will preserver the stacktrace

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+        setattr(obj, attr, wrapper)
+        error_occurred = False
+        try:
+            # This will show the detailed assert message.
```
Can you explain more about how this will show the detailed assert message? I don't see the connection at the outset and am not sure why it changes to `self.longMessage = false` later on. Not saying that it's wrong but it may need a bit more explanation.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+            # This will show the detailed assert message.
+            self.longMessage = True
+            yield
+        except Exception as exception:
```
Also just to check ... does self.assertEqual actually raise an exception? Can/should we make the caught exception more specific rather than just Exception?

cuichenli replied:
We need to catch all the exceptions - because the exceptions may be from the tested function.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -194,6 +194,11 @@ def _Dynamic_Fetch(self, request, response): # pylint: disable=invalid-name
         self.response = response
 
 
+class IntendedException(Exception):
```
This is not needed, right?

cuichenli replied:
Removed

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
             setattr(obj, attr, original)
 
+    @contextlib.contextmanager
+    def swap_with_checks(
```
This is a complex function. It's probably worth having some small unit tests in test_utils_test.py to check its various behaviours.

cuichenli replied:
Not sure how to unit test context manager. Any suggestions or example? Thanks!

cuichenli replied:
Added a few unit tests

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+            self.longMessage = False
+            setattr(obj, attr, original)
+            # Only do the rest of the checks when there was no error.
+            # Otherwise error in this check will overwrite the error raised
```
Could we make it such that the errors are aggregated? That would probably be more robust, rather than using "if" statements for fear of overwriting.

cuichenli replied:
This was added to avoid the `called_times` error overwrite the other ones, as that functionality has been removed, I will aggregate it.

seanlip replied:
Ah -- that actually raises another question. Suppose there are three elements in args, kwargs -- and the function is only called twice. Is that a problem? In any case we should document how that case is handled (and there should be unit tests too).

cuichenli replied:
This will be checked at the final part - whether the expected_args and expected_kwargs are empty.

> Could we make it such that the errors are aggregated? That would probably be more robust, rather than using "if" statements for fear of overwriting.

Also, I just realized it may be better to keep those separate. Otherwise, the error threw in the tested functions can not be tracked.

seanlip commented on scripts/common.py:
```diff
@@ -67,7 +73,8 @@ def is_linux_os():
     """Check if the architecture is on X64."""
-    return ARCHITECTURE == 'x86_64'
+    # https://docs.python.org/2/library/platform.html#platform.architecture
+    return sys.maxsize > 2**32  # pragma: no cover
```
You can test this. Use self.swap(sys, 'maxsize') and check both cases.

cuichenli replied:
Updated

seanlip commented on scripts/common.py:
```diff
@@ -437,6 +468,30 @@ def convert_to_posixpath(file_path):
+    to replace the file's content line by line.
+
+    Note:
+        This function was not designed to achive a comprehensive task. Instead,
```
Hm -- because of its non-specificity, this just inspires fear in developers :) 

Instead, use the docstring above to explain specifically and concretely what this function can be used for and what its specific limitations are (e.g. doesn't handle multiline regexes). That should suffice and provide the necessary guardrails for developers.

cuichenli replied:
Updated

seanlip commented on scripts/common.py:
```diff
@@ -437,6 +468,30 @@ def convert_to_posixpath(file_path):
 
 
+def inplace_replace_file(filename, regex_pattern, replace):
+    """Replace the file content inplace with regex pattern. The pattern is used
```
inplace --> in-place

cuichenli replied:
Updated

seanlip commented on scripts/common.py:
```diff
@@ -437,6 +468,30 @@ def convert_to_posixpath(file_path):
+    Args:
+        filename: str. The name of the file to be changed.
+        regex_pattern: str. The pattern to check.
+        replace: str. The content to be replaced.
```
replace --> replacement_string

(replace is a verb so it doesn't sound like a variable name/descriptor)

cuichenli replied:
Updated

seanlip commented on scripts/common.py:
```diff
@@ -437,6 +468,30 @@ def convert_to_posixpath(file_path):
+    regex = re.compile(regex_pattern)
+    for line in fileinput.input(
+            files=[filename], inplace=True, backup='.bak'):
+        # As PRINT is going to add a new line symbol, if we do not remove the
```
Hm, if this is done just for the PRINT, then we should just do the next three lines as:

```
    python_utils.PRINT(regex.sub(replace, line).rstrip())
```

right?

Also, I'm a bit confused because I can't see where the line gets written back to the file. Could you explain how this works?

cuichenli replied:
Aha, I know the `fileinput` API is really confusing, I got the same confusion. Basically all the things sent to stdout will be written back to the file. 

https://docs.python.org/2.7/library/fileinput.html
You can have a look at the `Optional in-place filtering` part.

seanlip replied:
Wow, that _is_ confusing.

Can we just not use it? Let's just do the "boring" thing of: "read the string from file, modify the string, write string back into file". Probably better to emphasize readability/understandability here.

cuichenli replied:
Changed

seanlip commented on scripts/common_test.py:
```diff
@@ -44,6 +47,29 @@
+    ]
+
+    def __init__(self, index):
+        self.index = index
```
Add a Args docstring for "index".

cuichenli replied:
Updated

seanlip reviewed (changes requested):
Thanks @cuichenli, I have reviewed up to MockPsutilProcess. Will continue later!

cuichenli committed:
Refine and simplify the `swap_with_check`

cuichenli committed:
Add unit tests for is_x64_architecture

cuichenli committed:
Add docstring for MockPsutilProcess

seanlip commented on scripts/install_chrome_on_travis_test.py:
```diff
@@ -0,0 +1,87 @@
+class InstallChromeOnTravisTests(test_utils.GenericTestBase):
+
+    def test_main(self):
+        chrome_source_url = 'https://github.com/webnicer/chrome-downloads/raw/master/x64.deb/google-chrome-stable_77.0.3865.75-1_amd64.deb' # pylint:disable=line-too-long
```
You don't need the pragma. Do something like this:

```
    chrome_source_url = (
        'https://github.com/webnicer/chrome-downloads/raw/master/x64.deb'
        '/google-chrome-stable_77.0.3865.75-1_amd64.deb')
```

cuichenli replied:
Updated

seanlip commented on scripts/install_chrome_on_travis_test.py:
```diff
@@ -0,0 +1,87 @@
+
+        def mock_chdir(unused_path):
+            return
+        # pylint: disable=unused-argument
```
Add newline above this one.

cuichenli replied:
Updated

seanlip commented on scripts/install_chrome_on_travis_test.py:
```diff
@@ -0,0 +1,87 @@
+        # pylint: disable=unused-argument
+        def mock_url_retrieve(unused_url, filename):
+            return
+        # pylint: enable=unused-argument
```
Add newline above this one.

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+         'core/tests/protractor/accessibility.js use --suite=accessibility.'
+         'For performing a full test, no argument is required.')
+
+SUBPROCESSES = []
```
Add short comment above this explaining what's stored in this array (and what their types are).

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+def check_screenshot():
+    """Check if screenshot directory exists, if so, delete it."""
+    screeenshots_dir = os.path.join(os.pardir, 'protractor-screenshots')
```
screenshots_dir not screeenshots_dir

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+        'Note: If ADD_SCREENSHOT_REPORTER is set to true in'
+        'core/tests/protractor.conf.js, you can view screenshots'
+        'of the failed tests in ../protractor-screenshots/')
+    os.rmdir(screeenshots_dir)
```
I think I'm slightly confused by this. The message above says "you can view stuff in this dir" ... and then we just proceed to delete the dir?

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+SUBPROCESSES = []
+
+
+def check_screenshot():
```
From reading the function, sounds like a better name would be something like ensure_screenshots_dir_is_removed().

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+
+def check_running_instance(*ports):
+    """Check if exists running oppia instance.
```
Check if a running Oppia instance exists on any of the given ports.

cuichenli replied:
Updated based on the below comment

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+        ports: list(int). A list of ports that oppia instances may be running.
+
+    Return:
+        bool: Whether the ports are all open.
```
Whether any port has a running Oppia instance. (perhaps?)

That said, I'm a bit confused, because the docstring talks about an "oppia instance" but it looks like this is only checking for whether ports are used in general, in a way that's agnostic to what process is being run on that port. Could the docstring be clarified?

Also the original return value description seems to have a confusion between "all ports are open" (based on the description) and "at least one port is open" (based on the impl). Not sure what is intended here.

cuichenli replied:
Updated based on the below comment

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+        common.kill_processes_based_on_regex(p)
+
+
+def check_running_instance(*ports):
```
is_any_port_open(port_numbers) perhaps.

(Distinguish between port and port number.)

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    return running
+
+
+def wait_for_port(port):
```
wait_for_port_to_be_open(port_number)

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    """
+    waited_seconds = 0
+    while (not common.is_port_open(port) and
+           waited_seconds < SECONDS_TO_WAIT_PORT):
```
SECONDS_TO_WAIT_PORT --> MAX_WAIT_TIME_FOR_PORT_TO_OPEN_SECS

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+        sys.exit(1)
+
+
+def tweak_constant_file(constant_file, dev_mode):
```
update_dev_mode_in_constants_js

dev_mode --> dev_mode_setting

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    common.inplace_replace_file(constant_file, pattern, replace)
+
+
+def tweak_feconf_file(feconf_file_path, enable_community_dashboard):
```
update_community_dashboard_status_in_feconf_file

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    """Build the javascript files.
+
+    Args:
+        dev_mode: bool. Represents whether run the related commands in dev
```
whether run --> whether to run

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+        install_chrome_on_travis.main(args=[])
+
+
+def build_js_files(dev_mode):
```
dev_mode --> dev_mode_setting

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+def tweak_webdriver_manager():
+    """webdriver-manager (version 13.0.0) uses `os.arch()` to determine the
+    architecture of the operation system, however, this function can only be
```
operation --> operating

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    """webdriver-manager (version 13.0.0) uses `os.arch()` to determine the
+    architecture of the operation system, however, this function can only be
+    used to determine the architecture of the machine that compiled `node`
+    (great job!). In the case of Windows, we are using the portable version,
```
I smiled a bit while reading this, but I think we might want to take some of the more personal bits out. Suggest eliding "(great job!)", and also changing "While clearly ..." to "Unfortunately, webdriver-manager seems to assume that Windows wouldn't run on the ia32 architecture, so its help function returns null for this, which means that the application has no idea about where to download the correct version."

What do you mean by "its own help function will return null for this", btw? Is os.arch() the "help function" you're referring to or is it something else?

Also, thank you for taking the time to explain clearly what's going on here and provide the links -- it helps a lot!

cuichenli replied:
Updated

The help function is the one decides the download link based on the architecture.

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+def start_webdriver_manager():
+    """Update and start webdriver manager."""
+    if common.is_windows_os():
+        tweak_webdriver_manager()
```
Instead of calling two completely separate tweak/untweak functions, would it be possible to use a "with" construct to swap the files out temporarily? That would be cleaner and easier to follow.

cuichenli replied:
I was thinking the same thing but not sure how to unit test a context manager.

seanlip replied:
There are a bunch of context managers in core/tests/test_utils.py, with tests in test_utils_test.py. Maybe you can look at those for an example? (Basically I think you can just write something that makes use of the context manager.)

cuichenli replied:
Added a few unit tests

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    the test is running on browserstack.
+
+    Args:
+        run_on_browserstack: bool. Whether run the test on browserstack.
```
Whether the test is being run on BrowserStack.

cuichenli replied:
Function removed

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+def get_parameter_for_config_file(run_on_browserstack):
+    """Return the parameter for the target configuration file based on whether
+    the test is running on browserstack.
```
BrowserStack

cuichenli replied:
Function removed

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+
+def get_parameter_for_sharding(sharding, sharding_instances):
+    """Return the parameter for sharding. Based on the sharding instances
```
sharding. Based --> sharding, based on the given number of sharding instances.

(don't forget the period at the end)

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    """Return the parameter for sharding. Based on the sharding instances
+
+    Args:
+        sharding: bool. Whether run the test with sharding.
```
run --> to run

That said ... isn't this kinda redundant with the second parameter? I mean we could just pass sharding_instances == 1.

Also the function should check that sharding_instances > 0 and raise an error if that does not hold.

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+                '--capabilities.maxInstances=%s' % sharding_instances]
+
+
+def get_parameter_for_dev_mode(dev_mode):
```
dev_mode --> dev_mode_setting

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+
+def get_parameter_for_dev_mode(dev_mode):
+    """Return parameter for whether the test should be running on dev_mode
```
Add period at end of sentence.

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+
+def get_parameter_for_suite(suite):
+    """Return parameter for which suite to be running on the tests.
```
Return a parameter for which suite to run the tests for.

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    """Return parameter for which suite to be running on the tests.
+
+    Args:
+        suite: str. The suite name to be running on the tests. If the value is
```
The suite name whose tests should be run. If the ...

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+            `full`, all tests will run.
+
+    Returns:
+        list(str): A list of command line parameter for suite.
```
parameter --> parameters

suite --> the suite

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    Args:
+        run_on_browserstack: bool. Represents whether the tests should run
+            on browserstack.
+        sharding: bool. Disables/Enables parallelization of protractor tests.
```
This argument could probably be omitted, see comment above.

cuichenli replied:
Removed

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+        sharding: bool. Disables/Enables parallelization of protractor tests.
+        sharding_instances: str. Sets the number of parallel browsers to open
+            while sharding.
+        suite: str. Performs test for different suites.
```
The name of the suite whose tests should be run.

(Also, maybe change this and similar args to suite_name.)

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+        sharding_instances: str. Sets the number of parallel browsers to open
+            while sharding.
+        suite: str. Performs test for different suites.
+        dev_mode: bool. Represents whether run the related commands in dev
```
dev_mode_setting

run --> to run

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    return commands
+
+
+def start_google_engine(dev_mode):
```
start_google_app_engine_server

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+
+def start_google_engine(dev_mode):
+    """Start the google engine server.
```
google engine --> Google App Engine

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    """Start the google engine server.
+
+    Args:
+        dev_mode: bool. Represents whether run the related commands in dev
```
dev_mode --> dev_mode_setting

Represents whether --> Whether to

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+    """Run the scripts to start end-to-end tests."""
+
+    parsed_args = _PARSER.parse_args(args=args)
+    other_instance_running = check_running_instance(
```
oppia_instance_is_already_running

cuichenli replied:
Updated

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,472 @@
+
+    parsed_args = _PARSER.parse_args(args=args)
+    other_instance_running = check_running_instance(
+        OPPIA_SERVER_PORT, GOOGLE_APP_ENGINE_PORT)
```
I think you can move these into the check_running_instance function if they're going to be the same every time. No need to pass them.

That way, you can actually make that function more oppia-specific and call it is_oppia_server_already_running().

cuichenli replied:
Updated

seanlip reviewed (changes requested):
Did a full pass. Thanks @cuichenli, it's awesome that this can be made to run on Windows too! 

Please don't hesitate to let me know if you have any questions about the comments I left.

seanlip commented:
Hi @cuichenli -- also, FYI, I checked with @kevinlee12 and he confirmed that we are not using BrowserStack. So I think you can get rid of all the browserstack stuff, which should hopefully make this script look a bit simpler.

Thanks!

cuichenli committed:
Rewrite inplace replace

cuichenli committed:
Add except part back

cuichenli committed:
Fix format

cuichenli committed:
Add comment for subprocesses

cuichenli committed:
Rename check_screenshot

cuichenli committed:
Rename check_running_instance

cuichenli committed:
Rename wait_for_port

cuichenli committed:
Rename tweak_const_js

cuichenli committed:
Rename tweak_feconf_file

cuichenli committed:
Rename and change comment

cuichenli committed:
Remove  browserstack related parts

cuichenli committed:
Remove the sharding part

cuichenli committed:
Rename a bunch of parameters

cuichenli committed:
Rename start_google_engine

cuichenli committed:
Rename is_port_open

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Fix linter issue

cuichenli committed:
Update comment

cuichenli committed:
Refine swap

cuichenli committed:
Refine sharding parameter function

cuichenli committed:
Fix linter issue

seanlip commented:
@cuichenli I think you missed this comment: https://github.com/oppia/oppia/pull/8159#discussion_r365503628

Also, the CircleCI/TravisCI checks are failing on this PR. Could you please look into that before I do the full review?

Thanks!

cuichenli commented:
@seanlip I can not reply to that comment somehow, but I made the change. I will have a look at the failure of the e2e test.

seanlip reviewed (commented):
FYI, I replied to https://github.com/oppia/oppia/pull/8159#discussion_r366637628

seanlip commented:
Ah - for replying to old comments, sometimes you might need to click the "5 days ago" link to get to the original comment thread that you can reply to. (Or find it by expanding the "XXX hidden items"  section in the comment thread.)

cuichenli committed:
Fix missed config file parameter

cuichenli committed:
Rewrite tweak_webdriver_file with context manager

cuichenli committed:
Add unit test for swap_with_check and refine the function

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

seanlip commented:
Hi @cuichenli, I'm trying to find my previous comments so that I can review the diff, but all the previous comments seem to have disappeared :( Did you force push or something?

seanlip commented:
(I can't even see your responses to those comments...)

seanlip commented:
Ah sorry, scratch that. Refreshing the page fixed it. Must be a GitHub bug.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1911,14 +1912,22 @@ def wrapper(*args, **kwargs):
             yield
+        except Exception:
+            error_occurred = True
+            # Raise issues threw by the called function or assert error.
```
threw --> thrown

cuichenli replied:
Done

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -161,17 +153,15 @@ def cleanup():
-    Args:
-        ports: list(int). A list of ports that oppia instances may be running.
+def is_oppia_server_already_running():
+    """Check if the ports are taken by any other processes. If the any one of
```
the any --> any

cuichenli replied:
Done

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -313,116 +312,91 @@ def undo_webdriver_tweak():
-    the test is running on browserstack.
+def get_parameter_for_sharding(sharding_instances):
+    """Return the parameter for sharding, based on the given number of
+        sharding instances.
```
Deindent by 4.

cuichenli replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -468,28 +468,32 @@ def convert_to_posixpath(file_path):
-def inplace_replace_file(filename, regex_pattern, replace):
-    """Replace the file content inplace with regex pattern. The pattern is used
+def inplace_replace_file(filename, regex_pattern, replacement_string):
+    """Replace the file content inp-lace with regex pattern. The pattern is used
```
inp-lace --> in-place

cuichenli replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -468,28 +468,32 @@ def convert_to_posixpath(file_path):
 
     Note:
-        This function was not designed to achive a comprehensive task. Instead,
+        This function was not designed to achieve a comprehensive task. Instead,
```
This note is still too vague. Please make it clearer. You can say "This function should only be used with files that are processed line by line." and that suffices (right?)

Otherwise developers will not know whether their task is "trivial" or "comprehensive". It's really important to be more specific here.

cuichenli replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -468,28 +468,32 @@ def convert_to_posixpath(file_path):
-    os.remove('%s.bak' % filename)
+    with python_utils.open_file(backup_filename, 'r') as f:
+        for line in f:
+            new_contents.append(regex.sub(replacement_string, line))
```
If some error arises in regex.sub or others, we still need to remove the backup file. So maybe use a try/finally. This case should be tested by swapping out one of the interior functions to raise an error, and we should verify that the backup file still gets deleted.

cuichenli replied:
Added a try and except. If there was any error, the except part will restore the files.

seanlip reviewed (changes requested):
Just a few more small notes, but generally, no further large concerns (the most important one from my point of view is the inplace/replace file docstring). Thanks @cuichenli!

seanlip commented:
@nithusha21 @vojtechjelinek PTAL as codeowners. Thanks!

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into convert-e2e-to-python

cuichenli committed:
Fix a few comments

cuichenli committed:
Improve in-place replace and add unit test

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

seanlip reviewed (approved):
LGTM as codeowner, thanks @cuichenli! PTAL at the CI tests though, they seem to be failing.

seanlip commented:
@nithusha21 PTAL

nithusha21 reviewed (approved):
LGTM for codeowner files!

nithusha21 commented:
Please note the merge conflicts and failing tests!

seanlip commented:
Per discussion with @cuichenli, I will be taking over this PR. I have created a replacement PR at  #8543 that supersedes this one.
